### ISSUE TITLE
CamembertForTokenClassification model always predict the same label for a multi-classification problem

### ISSUE THREAD

**Issue by vicoof** [pytorch/fairseq#1988]
Hi every one ! 

I'm working with the pretrained CamembertForTokenClassification  model for NER task and I don't undersatnd why I got the same logits and so the same labels batch after batch. The model doesn't learn and I don't know why. Do you have a clue ? 

**Code** : 

```
def train(train_dataloader,validation_dataloader,num_epochs:int):
    """"
    Train phase. Train + Validation steps. Print train loss and flat accuracy epochs after epochs.
    ===================================================
    params: 
        - train_dataloader : Data Loader of tensors train inputs, train labels and train masks.
        - num_epochs : Number of epochs for the train phase
        - num_batch : Size of the train batches
        
    returns : 
            /
    """
    # Function to calculate the accuracy of our predictions vs labels

    def flat_accuracy(preds, labels):
        pred_flat = np.argmax(preds, axis=2).flatten()
        labels_flat = labels.flatten()
        return np.sum(pred_flat == labels_flat) / len(labels_flat)
    
    # Camembert fine-tuning parameters
    param_optimizer = list(model.named_parameters())
    no_decay = ['bias', 'gamma', 'beta']
    optimizer_grouped_parameters = [
        {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)],
         'weight_decay_rate': 0.01},
        {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)],
         'weight_decay_rate': 0.0}
    ]

    optimizer = AdamW(model.parameters(), lr=1e-3, correct_bias=False)



    # Store our loss and accuracy for plotting
    train_loss_set = []

    # Number of training epochs 
    epochs = num_epochs

    # Camembert training loop
    for _ in trange(epochs, desc="Epoch"):  

        ## TRAINING

        # Set our model to training mode
        model.train()  
        # Tracking variables
        tr_loss = 0
        nb_tr_examples, nb_tr_steps = 0, 0
      # Train the data for one epoch
        for step, batch in enumerate(train_dataloader):

            # Add batch to GPU
            batch = tuple(t.to(device) for t in batch)
            # Unpack the inputs from our dataloader
            b_input_ids, b_input_mask, b_labels = batch
            # Clear out the gradients (by default they accumulate)
            optimizer.zero_grad()
            # Forward pass
            loss = model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask, labels=b_labels)[0]
            train_loss_set.append(loss.item())    
            # Backward pass
            loss.backward()
            # Update parameters and take a step using the computed gradient
            optimizer.step()
            # Update tracking variables
            tr_loss += loss.item()
            nb_tr_examples += b_input_ids.size(0)
            nb_tr_steps += 1
            print("Train loss: {}".format(tr_loss/nb_tr_steps))

      ## VALIDATION

      # Put model in evaluation mode
        model.eval()
      # Tracking variables 
        eval_loss, eval_accuracy = 0, 0
        nb_eval_steps, nb_eval_examples = 0, 0
        # Evaluate data for one epoch
        for batch in validation_dataloader:
            # Add batch to GPU
            batch = tuple(t.to(device) for t in batch)
            # Unpack the inputs from our dataloader
            b_input_ids, b_input_mask, b_labels = batch   
            # print('val_labels:',b_labels)     
            # Telling the model not to compute or store gradients, saving memory and speeding up validation
            with torch.no_grad():
              # Forward pass, calculate logit predictions
              logits = model(b_input_ids, token_type_ids=None, attention_mask=b_input_mask)[0]
            # Move logits and labels to CPU
            logits = logits.detach().cpu().numpy()
            print('logits :',logits)
            print('labels predicted:',np.argmax(logits, axis=2).flatten())
            label_ids = b_labels.to('cpu').numpy()
            # print('rela labels:',label_ids)
            tmp_eval_accuracy = flat_accuracy(logits, label_ids) 
            eval_accuracy += tmp_eval_accuracy
            nb_eval_steps += 1
        print("Validation Accuracy: {}".format(eval_accuracy/nb_eval_steps))


    # plot training performance
    plt.figure(figsize=(15,8))
    plt.title("Training loss")
    plt.xlabel("Batch")
    plt.ylabel("Loss")
    plt.plot(train_loss_set)
    plt.show()

```

**After few batches,  here's what I got during a validation phase for a specific batch:** 

```

predicted logits : [[[-1.0357289   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099023  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357289   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  ...
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]]

 [[-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169782 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169782 -0.13911003 -1.0021709 ]
  ...
  [-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]]

 [[-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  [-1.0357288   1.209902   -0.30169785 -0.13911004 -1.0021709 ]
  ...
  [-1.0357288   1.2099023  -0.30169782 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]]

 ...

 [[-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  ...
  [-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169782 -0.13911004 -1.0021708 ]
  [-1.0357288   1.2099022  -0.30169782 -0.13911004 -1.0021709 ]]

 [[-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169782 -0.13911004 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911004 -1.0021709 ]
  ...
  [-1.0357288   1.2099022  -0.30169782 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911003 -1.0021709 ]
  [-1.0357288   1.2099023  -0.30169785 -0.13911006 -1.0021709 ]]

 [[-1.0357288   1.2099023  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  ...
  [-1.0357289   1.2099022  -0.30169785 -0.13911007 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]
  [-1.0357288   1.2099022  -0.30169785 -0.13911006 -1.0021709 ]]]
labels predicted: [[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]]
real validation labels: tensor([[1, 2, 2, 2, 1, 1, 2, 2, 2, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0],
        [2, 2, 2, 2, 1, 1, 2, 2, 1, 1, 2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0],
        [1, 1, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 2, 2, 1, 1, 1, 1, 1, 1, 3, 3, 3, 3,
         2],
        [2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 1, 1,
         1],
        [4, 4, 1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 1, 0, 0, 0,
         0],
        [4, 4, 4, 4, 4, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1],
        [1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2,
         2],
        [2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 1, 1, 1, 1, 4, 4, 1, 1, 0, 0, 0, 0, 0,
         0],
        [3, 3, 3, 3, 3, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0,
         0],
        [1, 1, 1, 1, 1, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
         0]], device='cuda:0')

```

**Model** : 

`model = CamembertForTokenClassification.from_pretrained('camembert base',num_labels=num_labels).cuda()`

**Model config :** 

```

CamembertConfig {
  "_num_labels": 5,
  "architectures": [
    "CamembertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "bad_words_ids": null,
  "bos_token_id": 5,
  "decoder_start_token_id": null,
  "do_sample": false,
  "early_stopping": false,
  "eos_token_id": 6,
  "finetuning_task": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LABEL_4"
  },
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "is_decoder": false,
  "is_encoder_decoder": false,
  "label2id": {
    "LABEL_0": 0,
    "LABEL_1": 1,
    "LABEL_2": 2,
    "LABEL_3": 3,
    "LABEL_4": 4
  },
  "layer_norm_eps": 1e-05,
  "length_penalty": 1.0,
  "max_length": 20,
  "max_position_embeddings": 514,
  "min_length": 0,
  "model_type": "camembert",
  "no_repeat_ngram_size": 0,
  "num_attention_heads": 12,
  "num_beams": 1,
  "num_hidden_layers": 12,
  "num_return_sequences": 1,
  "output_attentions": false,
  "output_hidden_states": false,
  "output_past": true,
  "pad_token_id": 1,
  "prefix": null,
  "pruned_heads": {},
  "repetition_penalty": 1.0,
  "task_specific_params": null,
  "temperature": 1.0,
  "top_k": 50,
  "top_p": 1.0,
  "torchscript": false,
  "type_vocab_size": 1,
  "use_bfloat16": false,
  "vocab_size": 32005
}
```
> 

**What I tried :** 

I tried to tweak several hyperparameter like the hidden_dropout_prob ratio, the learning rate of the sizes of the batches and padding length but nothing worked...

**My environment**`

 - fairseq Version (e.g., 1.0 or master):
 - PyTorch Version : 1.4.0
 - OS (e.g., Linux): Windows 10
 - Python version: 3.6.9
 - transformers: 2.8.0
 - GPU models and configuration:
_CudaDeviceProperties(name='Tesla K80', major=3, minor=7, total_memory=11441MB, multi_processor_count=13)

Thank you for your help :)

**Comment by huihuifan** [comment]
@louismartin

**Comment by vicoof** [comment]
Hi Louis,

Thanks for answering me. Could you be a little more explicit please ? I’m not very familiar with Bert...

Best,

**Comment by louismartin** [comment]
Hi, sorry for the delay. 
As I mentioned in my email, you should check if the weights of the model change during training.
Do you see the training loss decreasing or is it constant ?

**Comment by vicoof** [comment]
Hi no worry. 
The weights are actually changing during training and the loss is decreasing untill a certain point.

**Comment by louismartin** [comment]
And you confirm that the logits are exactly the same at the beginning of training and at the end of training when the weights have changed?

**Comment by louismartin** [comment]
You should try to follow the path from the input embeddings to the output embeddings and try to see what happens. If the weights are different, then the output should be different.

**Comment by vicoof** [comment]
> And you confirm that the logits are exactly the same at the beginning of training and at the end of training when the weights have changed?

Yes the logits are evolving but converge towards a positive value for only one class and negative values for all the others. So the same class si predicted when using the argmax of the logits.

**Comment by louismartin** [comment]
Then I think it might be related to your training data or the way you feed the data/labels. I can't help you more on this...
Make sure that you don't always feed the same label during training. Maybe one label is way more frequent than the others. You could investigate what happens when you remove this label from training.

**Comment by vicoof** [comment]
Indeed, one label is way more present than the others but when I remove it, the model predicts another one all the time...
Thanks anyway,

**Comment by louismartin** [comment]
Maybe you could try on the same datasets that we used in the paper.
If you take the french part of the XNLI dataset, you can finetune using these instructions (MNLI part) .
You could also create a very simple dummy synthetic dataset to check that the training procedure works well. For instance a dataset where the label is the number of words or something like this.

**Comment by vicoof** [comment]
Hi,
Looks like I messed with the labeling phase. I didn't understand well the mask and the padding goal. In fact, I added the label 0 for 'Unused label' which should be the label for the padded input. I didn't know that Bert was taking care of this already.
Anyway, it's working fine now...

**Comment by vicoof** [comment]
Hi,
Looks like I messed up with the labeling phase. I didn't understand well the mask and the padding goal. In fact, I added the id label 0 for 'Unused input' which should be the label for the padding inputs. By doing this, I added a label which was not considered for my initial NER task ( like PERS, LOC or ORG) . So the algorithm abviously didn't work well. I didn't know that Bert was already taking care of this .
Anyway, it's working fine now...
Sorry for the trouble and thank you for your help.

**Comment by ameyaparab28** [comment]
> Hi,
> Looks like I messed up with the labeling phase. I didn't understand well the mask and the padding goal. In fact, I added the id label 0 for 'Unused input' which should be the label for the padding inputs. By doing this, I added a label which was not considered for my initial NER task ( like PERS, LOC or ORG) . So the algorithm abviously didn't work well. I didn't know that Bert was already taking care of this .
> Anyway, it's working fine now...
> Sorry for the trouble and thank you for your help.

Hello,

Can you please elaborate.on your fix? I'm facing a similar issue with CamemBertForSequenceClassification, wherein the model is predicting the same class over and over again for the entire dataset. The exact same dataset (English) version works perfectly with the Bert model (BertForSequenceClassification), but the translated French text predicts the same class for all of the examples with CamemBert.
